### PR TITLE
convert to class

### PR DESCRIPTION
**Pull request by SparshithNR**
1. Update to class 
2. ~Rewrite test suite with broccoli-test-helpers~ moved to separate [PR](https://github.com/broccolijs/broccoli-merge-trees/pull/73)

### CODE REVIEW

**Comment by rwjblue** [comment]
Ready for a rebase

File: index.js [hunk 1]
```diff
@@ -3,44 +3,47 @@
...
-      overwrite: this.options.overwrite,
-      annotation: this.options.annotation
+class BroccoliMergeTrees extends Plugin{
+  constructor (inputNodes, options = {}) {
+    super(inputNodes, {
+      persistentOutput: true,
+      needsCache: false,
+      annotation: options.annotation
     });
+    var name = 'broccoli-merge-trees:' + (options.annotation || '');
```
Comment (rwjblue) [inline index.js]: Swap to `let`?

File: index.js [hunk 2]
```diff
@@ -3,44 +3,47 @@
...
+    if (!Array.isArray(inputNodes)) {
```
Comment (rwjblue) [inline index.js]: Can you move this error above the `super` call?

File: index.js [hunk 3]
```diff
@@ -3,44 +3,47 @@
...
+
+        err.message = `${message}\n${err.message}`;
+      }
+      throw err;
     }
-    throw err;
   }
+}
+
+module.exports = function concatPlugin(...params) {
```
Comment (rwjblue) [inline index.js]: This seems wrong, I don't believe that `concatPlugin` is the right name for this func

File: test.js [hunk 4]
```diff
@@ -1,49 +1,72 @@
 'use strict';
 
 var MergeTrees = require('./');
+const { createBuilder, createTempDir } = require('broccoli-test-helper');
```
Comment (rwjblue) [inline test.js]: Please split the test changes into a separate PR, I would like to land the class update separately

File: index.js (hunk 3 above)
Comment (SparshithNR) [reply index.js]: That's my miss. Updated the PR

File: test.js (hunk 4 above)
Comment (SparshithNR) [reply test.js]: Done #73

File: index.js [hunk 5]
```diff
@@ -1,46 +1,51 @@
...
+    if (!Array.isArray(inputNodes)) {
+      throw new TypeError(name + ': Expected array, got: [' + inputNodes +']');
+    }
+    super(inputNodes, {
+      persistentOutput: true,
+      needsCache: false,
+      annotation: options.annotation
     });
+
+    let name = 'broccoli-merge-trees:' + (options.annotation || '');
```
Comment (rwjblue) [inline index.js]: Hmm, this seems like it should be above the `if (!Array.isArray` check (right at the beginning of the constructor) since `name` is used there for the error message.

File: index.js [hunk 6]
```diff
@@ -1,46 +1,51 @@
...
     }
-    throw err;
   }
+}
+
+module.exports = function mergeTrees(...params) {
+  return new BroccoliMergeTrees(...params);
 };
+
+module.exports.BroccoliMergeTrees = BroccoliMergeTrees;
```
Comment (rwjblue) [inline index.js]: Let's call it `MergeTrees`:

```suggestion
module.exports.MergeTrees = MergeTrees;
```

This way folks that do want the class usage can do:

```js
const { MergeTrees } = require('broccoli-merge-trees');
```

**Comment by stefanpenner** [comment]
rebased + fixed style issues, let see if tests pass

**Comment by stefanpenner** [comment]
@SparshithNR 

> Looking good, I think we need to update the README with some additional information:

> the primary mechanism is not via new
> that you can use the base class from require('broccoli-merge-trees').MergeTrees

Have you addressed rob's feedback ^^^ ?

**Comment by SparshithNR** [comment]
Yes. I have updated the README accordingly. Please review. 
@rwjblue @stefanpenner

File: index.js [hunk 7]
```diff
@@ -1,46 +1,51 @@
...
 
-BroccoliMergeTrees.prototype.build = function() {
-  if (this.mergeTrees == null) {
-    // Defer instantiation until the first build because we only
-    // have this.inputPaths and this.outputPath once we build.
-    this.mergeTrees = new MergeTrees(this.inputPaths, this.outputPath, {
-      overwrite: this.options.overwrite,
-      annotation: this.options.annotation
+class BroccoliMergeTrees extends Plugin {
+  constructor (inputNodes, options = {}) {
```
Comment (stefanpenner) [inline index.js]: extra whitespace

File: README.md [hunk 8]
```diff
@@ -12,11 +12,27 @@ npm install --save-dev broccoli-merge-trees
 ```
 
 ## Usage
-
+1. With `new`
 ```js
 var BroccoliMergeTrees = require('broccoli-merge-trees');
```
Comment (stefanpenner) [inline README.md]: let's drop `var` for `let` or `const`

File: README.md [hunk 9]
```diff
@@ -12,11 +12,27 @@ npm install --save-dev broccoli-merge-trees
 ```
 
 ## Usage
+1. With `new`
+```js
+const BroccoliMergeTrees = require('broccoli-merge-trees');
 
+let mergedNode = new BroccoliMergeTrees(inputNodes, options);
+```
```
Comment (rwjblue) [inline README.md]: This seems wrong, we shouldn't suggest using `new` with the plain function form. Cases 1 and 3 in this list should be collapsed to a single entry.

```js
const { MergeTrees } = require('broccoli-merge-trees');

let mergedNode = new MergeTrees(inputNodes, options);
```

File: README.md [hunk 10]
```diff
@@ -12,11 +12,27 @@ npm install --save-dev broccoli-merge-trees
...
+2. As a `function call`
 ```js
-var BroccoliMergeTrees = require('broccoli-merge-trees');
+const broccoliMergeTrees = require('broccoli-merge-trees');
+
+let mergedNode = broccoliMergeTrees(inputNodes, options);
+```
+
+3. As a `base class`
```
Comment (rwjblue) [inline README.md]: We do not support folks extending from our Plugin base class. We expect to only be used directly for the functionality we provide.

File: README.md [hunk 11]
```diff
@@ -37,7 +53,7 @@ var mergedNode = new BroccoliMergeTrees(inputNodes, options);
 If this is your `Brocfile.js`:
 
 ```js
-var BroccoliMergeTrees = require('broccoli-merge-trees');
+const BroccoliMergeTrees = require('broccoli-merge-trees');
 
 module.exports = new BroccoliMergeTrees(['public', 'scripts']);
```
Comment (rwjblue) [inline README.md]: This snippet is wrong:

* `Brocfile.js` supports exporting a function, just like `ember-cli-build.js`, we should update to
* we should not be recommending that folks use `new` with our main export (use `.MergeTrees` for that)

I believe it should be:

```js
const mergeTrees = require('broccoli-merge-trees');

module.exports = function() {
  return mergeTrees(['public','scripts']);
};
```

File: index.js [hunk 12]
```diff
@@ -1,46 +1,51 @@
...
+    if (!Array.isArray(inputNodes)) {
+      throw new TypeError(name + ': Expected array, got: [' + inputNodes +']');
+    }
+    super(inputNodes, {
+      persistentOutput: true,
+      needsCache: false,
+      annotation: options.annotation
     });
+
+
```
Comment (rwjblue) [inline index.js]: Probably don't need multiple newlines here

File: README.md [hunk 13]
```diff
@@ -12,11 +12,17 @@ npm install --save-dev broccoli-merge-trees
 ```
 
 ## Usage
+1. With `new`
```
Comment (stefanpenner) [inline README.md]: I believe @rwjblue's feedback still needs to be addressed: 

> the primary mechanism is not via new

**Comment by stefanpenner** [comment]
released as v4.0.0 🎉
